### ISSUE TITLE
[Help wanted] Default window decorations color

### ISSUE THREAD

**Issue by itspers** [microsoft/wslg#37]
Big chance its not WSLg issue, but maybe somebody will able to tell where to look:

Most apps respect current GTK_THEME and use **window decorations** from it. But some, most important JetBrains IDEs won't and for example xcalc.

What part of this magic system is responsible to choose what window decoration theme to use? Can i force it to use some dark theme by default?

<img width="1038" alt="win-decorations" src="https://user-images.githubusercontent.com/833192/115714487-d5ea2c80-a37f-11eb-847e-0525e99eaa35.png">

**Comment by kidcats** [comment]
when I use JetBrains IDEs, I find that there is no full screen model in view tool. Can you use the full screen model?

**Comment by itspers** [comment]
> when I use JetBrains IDEs, I find that there is no full screen model in view tool. Can you use the full screen model?

no, but it wasn't there in 'old' xserver mode as well, im not sure if its linux versions of Jetbrains apps or particular problem with WSL...

**Comment by kidcats** [comment]
> > when I use JetBrains IDEs, I find that there is no full screen model in view tool. Can you use the full screen model?
> 
> no, but it wasn't there in 'old' xserver mode as well, im not sure if its linux versions of Jetbrains apps or particular problem with WSL...

so sad .i hope jetbrains or microsoft can solve this.

**Comment by brneor** [comment]
@itspers how you managed to set Adwaita Dark as default theme?

**Comment by spronovo** [comment]
Thanks for the suggestion. Who paints the window decoration depends on the application. If the app let the window manager paint it, than it's the window manager inside of Weston that paints it (the right side window). We're currently just using the default theme, was was good enough for the preview :-)... we would like to eventually add a theme that more closely resemble Windows itself and to synchronize it with your Windows themes, but that is some ways out.

Other application can decide to paint those area themselves instead (the left case)... in that case any system level themes will have no effect and instead this is controlled entirely by the application itself (or the framework it uses).

**Comment by itspers** [comment]
>we would like to eventually add a theme that more closely resemble Windows itself and to synchronize it with your Windows themes, but that is some ways out.

This will be also not ideal solution. Vcxsrv/x410/gWsl using Windows native decorations in such cases without CSD and it results to same white window headers if you using light windows theme. 

@spronovo is it theoretically possible to ssh somehow to running distro and install-switch theme? Or it not worth trying and need only to make own build from sources?

**Comment by spronovo** [comment]
The system distro is mounted as read-only... so any changes you make it to are discarded when WSL is restarted. The servicing model we have for the system distro is a simple throwaway and replace with the new one... if we allowed you to persist data in it you would be loosing it on update. If you want a truly custom experience, building your own system distro is one way to go... but you'll need to keep rebuilding as we provide updates if you want to benefit from those fixes and improvement.

**Comment by NyxCode** [comment]
Apart from theming, which would be super nice, the window decoration does not integrate well into windows.
For example, I would have expected being able to drag it to the edge of the screen to resize it or drag it when the window is maximized to exit fullscreen.

Would it be possible to render the window decoration (unless the application renders its own) in windows, maybe through the RDP client? Wouldn't we get theming & better windows integration for "free" then?

**Comment by itspers** [comment]
> Apart from theming, which would be super nice, the window decoration does not integrate well into windows.
> For example, I would have expected being able to drag it to the edge of the screen to resize it or drag it when the window is maximized to exit fullscreen.
> 
> Would it be possible to render the window decoration (unless the application renders its own) in windows, maybe through the RDP client? Wouldn't we get theming & better windows integration for "free" then?

No, because https://github.com/microsoft/wslg/issues/37#issuecomment-826703442

It will be same as now on all existing xservers - or dark or light, without respect for application content at all

**Comment by lirannl** [comment]
How did you get your gtk apps to use a different gtk theme at all? I couldn't make a difference even though I set GTK_THEME in /etc/profile...

**Comment by erickvieira** [comment]
> How did you get your gtk apps to use a different gtk theme at all? I couldn't make a difference even though I set GTK_THEME in /etc/profile...

Via gnome tweaks, maybe.

$ `sudo apt install gnome-tweaks`
$ `gnome-tweaks &`

Then you can download a custom theme from [Gnome Look](https://www.gnome-look.org/browse/cat/) and select it on tweaks window.

**Comment by kidcats** [comment]
> > How did you get your gtk apps to use a different gtk theme at all? I couldn't make a difference even though I set GTK_THEME in /etc/profile...
> 
> Via gnome tweaks, maybe.
> 
> $ `sudo apt install gnome-tweaks`
> $ `gnome-tweaks &`
> 
> Then you can download a custom theme from [Gnome Look](https://www.gnome-look.org/browse/cat/) and select it on tweaks window.

but i can not select the theme in gnome-tweaks after installed gnome-tweaks,can you tell me how did you do that

**Comment by dcharlespyle** [comment]
Microsoft-Edge-Dev for Linux now is another one of the apps that no longer displays its window decorations like previous builds of the app once did.  Now, Edge for Linux no longer displays its own window decorations in Wayland in WSLg since its last build.

**Comment by onomatopellan** [comment]
VSCode also looks great when you change the titlebar style to custom

![image](https://user-images.githubusercontent.com/599142/124748710-bba60080-df23-11eb-9c6d-5fe466ff63ed.png)

**Comment by ei-eiron** [comment]
I'm not seeing a way to change the default Weston theme? Maintaining a custom build sounds like more work than squinting at the light theme.

User input:
- Inheriting Win11 theme is what I was expecting for Firefox, URXVT, or non-GTK/QT apps (but I am giddy with delight at the just-workingness of WSLg generally)
- Somehow listening to GTK settings would be ideal, esp `GTK_THEME`, but please don't throw out the first option in pursuit of the second

**Comment by ariaieboy** [comment]
please upvote this on youtrack
https://youtrack.jetbrains.com/issue/WI-65906

**Comment by delucca** [comment]
Anyone was able to find a workaround this? Or at least a way to use fullscreen for those applications (Jetbrains IDEs especially)

**Comment by MrStillwater** [comment]
For Chrome you can set 'Preferred Ozone platform' to 'Wayland' in chrome://flags
This might work for Edge as well.

**Comment by dcharlespyle** [comment]
> For Chrome you can set 'Preferred Ozone platform' to 'Wayland' in chrome://flags This might work for Edge as well.

Edge in Linux does not appear to have the option available.  But it works for Chrome.

**Comment by MrStillwater** [comment]
It seems like the long-term solution is for apps/environments to support Wayland directly.  Nautilus, gedit, etc. look great!

Here's the latest re: Jetbrains IDEs: https://youtrack.jetbrains.com/issue/JBR-3206

**Comment by xsrvmy** [comment]
FWIW, the problem is applications that don't have a CSD option rather than no wayland support. I assume VSCode wouldn't do this with CSD enabled even without ozone wayland.

**Comment by sharpjs** [comment]
How hard is it for WSLg to just match the (Windows) system theme?  There are going to be non-CSD apps for a long, long time.

**Comment by sarim** [comment]
I was researching about this issue a bit and found this article: Explains the situation really well. https://news.itsfoss.com/wayland-core-protocol-issue/

Technically WSLg-weston is doing Server Side Decorations (for non-gnome applications) right? Otherwise we wouldn't have any titlebar. Where that code comes from? Because from google it seems like weston doesn't have SSD. So WSLg's weston fork adds that functionality? Guidance about which file to look at for SSD Drawing code would be very helpful :3

**Comment by MamiyaOtaru** [comment]
You can change this by building your own version of the WSLg System Distro.  Instructions here: https://github.com/microsoft/wslg/blob/main/CONTRIBUTING.md

Once everything is downloaded, before compiling, you'll need to change a few files in wslg/vendors/weston/shared (cairo-util.c and frame.c)  Patch below (first chunk is frame color, second chunk is frame text color, third chunk is button color).  

Also edit the pngs for the titlebar buttons in wslg/vendors/weston/data (I just inverted them).  Compile your new System Distro and use it per the instructions at the link above

<img width="874" alt="wslgDarkMode" src="https://user-images.githubusercontent.com/2042826/219785856-ba96ad7a-7b9e-4896-84f1-cc066138fc5b.png">

diff -u ./shared/cairo-util.c ./dark/cairo-util.c
--- ./shared/cairo-util.c	2023-02-17 13:08:21.716641386 -0700
+++ ./dark/cairo-util.c	2023-02-13 15:30:32.492007835 -0700
@@ -372,12 +372,12 @@
 
 	if (flags & THEME_FRAME_ACTIVE) {
 		pattern = cairo_pattern_create_linear(16, 16, 16, 112);
-		cairo_pattern_add_color_stop_rgb(pattern, 0.0, 1.0, 1.0, 1.0);
-		cairo_pattern_add_color_stop_rgb(pattern, 0.2, 0.8, 0.8, 0.8);
+		cairo_pattern_add_color_stop_rgb(pattern, 0.0, 0.08, 0.09, 0.098);
+		cairo_pattern_add_color_stop_rgb(pattern, 0.2, 0.16, 0.175, 0.191);
 		cairo_set_source(cr, pattern);
 		cairo_pattern_destroy(pattern);
 	} else {
-		cairo_set_source_rgba(cr, 0.75, 0.75, 0.75, 1);
+		cairo_set_source_rgba(cr, 0.165, 0.18, 0.196, 1);
 	}
 }
 
@@ -567,14 +567,14 @@
 
 		if (flags & THEME_FRAME_ACTIVE) {
 			cairo_move_to(cr, x + 1, y  + 1);
-			cairo_set_source_rgb(cr, 1, 1, 1);
+			cairo_set_source_rgb(cr, 0, 0, 0);
 			SHOW_TEXT(cr);
 			cairo_move_to(cr, x, y);
-			cairo_set_source_rgb(cr, 0, 0, 0);
+			cairo_set_source_rgb(cr, 1, 1, 1);
 			SHOW_TEXT(cr);
 		} else {
 			cairo_move_to(cr, x, y);
-			cairo_set_source_rgb(cr, 0.4, 0.4, 0.4);
+			cairo_set_source_rgb(cr, 0.6, 0.6, 0.6);
 			SHOW_TEXT(cr);
 		}
 	}
diff -u ./shared/frame.c ./dark/frame.c
--- ./shared/frame.c	2023-02-17 13:08:08.496645466 -0700
+++ ./dark/frame.c	2023-02-13 15:22:09.211776901 -0700
@@ -232,11 +232,11 @@
 		cairo_stroke_preserve(cr);
 
 		if (button->press_count) {
-			cairo_set_source_rgb(cr, 0.7, 0.7, 0.7);
+			cairo_set_source_rgb(cr, 0.08, 0.09, 0.098);
 		} else if (button->hover_count) {
-			cairo_set_source_rgb(cr, 1.0, 1.0, 1.0);
+			cairo_set_source_rgb(cr, 0.21, 0.225, 0.241);
 		} else {
-			cairo_set_source_rgb(cr, 0.88, 0.88, 0.88);
+			cairo_set_source_rgb(cr, 0.16, 0.175, 0.191);
 		}
 
 		cairo_fill (cr);

**Comment by sharpjs** [comment]
Thank you, @MamiyaOtaru!  I had hoped the solution was something as simple as that.  Next step is for someone [smarter than me] to make it configurable, or even automatic.

**Comment by cazzoo** [comment]
I confirm this should be configuration (or should use windows color scheme automatically to detect)

**Comment by pingu6** [comment]
any update on this?

**Comment by sarim** [comment]
@pkkid technically XWayland, not the actual X11. as wslg is a wayland system. So at the end its weston that ends up drawing the window.

**Comment by diskree** [comment]
Isn't it possible to use a regular window frame in WSLg? I would like to see a frame with buttons like this

![image](https://github.com/microsoft/wslg/assets/96978370/a76d7912-21a7-4c63-8e7a-511f4ac7edb3)

**Comment by pscheit** [comment]
X410 works fine. I then ended up [VcXsrv](https://sourceforge.net/projects/vcxsrv/) which had less quirks with my software (PHPStorm from Jetbrains).
Others might have a different experience with it :)

**Comment by motabass** [comment]
On latest IntelliJ you can just use this VM Option to use Wayland instead:

`-Dawt.toolkit.name=WLToolkit`

**Comment by ivanocj** [comment]
How did you make it to work with weston? I've got:
`` 
java.awt.AWTError: Can't connect to the Wayland server
``

> On latest IntelliJ you can just use this VM Option to use Wayland instead:
> 
> `-Dawt.toolkit.name=WLToolkit`

**Comment by andrii0lomakin** [comment]
> @motabass How did you make it to work with weston? I've got: ` java.awt.AWTError: Can't connect to the Wayland server`
> 
> > On latest IntelliJ you can just use this VM Option to use Wayland instead:
> > `-Dawt.toolkit.name=WLToolkit`

Please fill issue in JBR YouTrack project that is a bug.

**Comment by ivanocj** [comment]
> 
> Hi, I had the same problem with Ubuntu 24.04. This was the solution for me:
> 
> `ln -sf /mnt/wslg/runtime-dir/wayland-* $XDG_RUNTIME_DIR/`
> 
> #1244

Thanks so much @eFzvai. I can confirm this symlink workaround has fixed Intellij to find Wayland on WSL2 (ubuntu 24.04)

**Comment by alexvorobiev** [comment]
Are there any updates on the issue? I use dark theme in both W11 and Linux GUI applications (emacs primarily) and the bright window borders look very out of place. Thanks!

**Comment by AngusYGChen** [comment]
> > [@motabass](https://github.com/motabass) How did you make it to work with weston? I've got: ` java.awt.AWTError: Can't connect to the Wayland server`
> > > On latest IntelliJ you can just use this VM Option to use Wayland instead:
> > > `-Dawt.toolkit.name=WLToolkit`
> 
> Hi, I had the same problem with Ubuntu 24.04. This was the solution for me:
> 
> `ln -sf /mnt/wslg/runtime-dir/wayland-* $XDG_RUNTIME_DIR/`
> 
> [#1244](https://github.com/microsoft/wslg/issues/1244)

thanks ! this is helpful for me !

**Comment by dcharlespyle** [comment]
> ln -sf /mnt/wslg/runtime-dir/wayland-* $XDG_RUNTIME_DIR/

Unfortunately, that doesn't work for me.  I get the following error when running the above command:

```
ln: '/mnt/wslg/runtime-dir/wayland-0' and '/mnt/wslg/runtime-dir/wayland-0' are the same file
ln: '/mnt/wslg/runtime-dir/wayland-0.lock' and '/mnt/wslg/runtime-dir/wayland-0.lock' are the same file
```

**Comment by ivanocj** [comment]
@dcharlespyle , looks like you have an outdated WSL.
`wsl --update`

Then you'll need to register a new ubuntu (>= 22.04).

**Comment by dcharlespyle** [comment]
> [@dcharlespyle](https://github.com/dcharlespyle) , looks like you have an outdated WSL2 (2.5.7 is the latest version today). `wsl --update`
> 
> Then you'll need to register a new ubuntu (>= 22.04).

![Image](https://github.com/user-attachments/assets/66844f63-bcbe-4ddd-bf0d-76a5c01be478)
As to the Linux version, that might need to be updated.

**Comment by qq1038765585** [comment]
You can try my modified wslg system distro, Can better adapt to light/dark mode：
https://github.com/qq1038765585/wslg_title_bar_beautify

![Image](https://github.com/user-attachments/assets/b2b24fec-94ef-48dd-9810-924fd47df564)

**Comment by shawny43** [comment]
Has i seen, it's QT programs with missing theme. If you can install the fedora package for qt titlebars this may solve your problem. qadwaitadecorations-qt5/qt6
